src/tests/Equinor.ProCoSys.Auth.Tests/Permission/MainApiPermissionServiceTests.cs

```diff
@@ -0,0 +1,211 @@
+            _mainApiClient = new Mock<IMainApiClient>();
+
+            _mainApiTokenProvider = new Mock<IMainApiAuthenticator>();
+            _dut = new MainApiPermissionService(_mainApiTokenProvider.Object, _mainApiClient.Object, _mainApiOptions.Object);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldReturnCorrectNumberOfPlants()
+        {
+            // Arange
+            _mainApiClient
+                .Setup(x => x.QueryAndDeserializeAsync<List<AccessablePlant>>(It.IsAny<string>(), null))
+                .Returns(Task.FromResult(new List<AccessablePlant>
+                {
+                    new AccessablePlant { Id = "PCS$ASGARD", Title = "Åsgard" },
+                    new AccessablePlant { Id = "PCS$ASGARD_A", Title = "ÅsgardA" },
+                    new AccessablePlant { Id = "PCS$ASGARD_B", Title = "ÅsgardB" },
+                }));
+
+            // Act
+            var result = await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            Assert.AreEqual(3, result.Count);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldSetsCorrectProperties()
+        {
+            // Arange
+            var proCoSysPlant = new AccessablePlant { Id = "PCS$ASGARD_B", Title = "ÅsgardB" };
+            _mainApiClient
+                .Setup(x => x.QueryAndDeserializeAsync<List<AccessablePlant>>(It.IsAny<string>(), null))
+                .Returns(Task.FromResult(new List<AccessablePlant>
+                {
+                    proCoSysPlant,
+                }));
+            // Act
+            var result = await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            var plant = result.Single();
+            Assert.AreEqual(proCoSysPlant.Id, plant.Id);
+            Assert.AreEqual(proCoSysPlant.Title, plant.Title);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldSetApplicationAuthentication()
+        {
+            // Act
+            await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            _mainApiTokenProvider.VerifySet(a => a.AuthenticationType = AuthenticationType.AsApplication);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldResetToOnBehalfOfAuthentication()
+        {
+            // Act
+            await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            _mainApiTokenProvider.VerifySet(a => a.AuthenticationType = AuthenticationType.OnBehalfOf);
+        }
+
+        [TestMethod]
+        public async Task GetPermissions_ShouldReturnThreePermissions_OnValidPlant()
+        {
+            // Arrange
+            _mainApiClient
+                .SetupSequence(x => x.QueryAndDeserializeAsync<List<string>>(It.IsAny<string>(), null))
+                .Returns(Task.FromResult(new List<string>{ "A", "B", "C" }));
```

Comment:
ReturnAsync for the rest of the file where appropriate

Reply:
ahh ... didn' know about it .. nice